[10/N] Update barrier with CPU/CUDA implementations

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #87088 Remove ProcessGroupRoundRobin
* #86409 [13/N] Update gather with CPU/CUDA implementations
* #86408 [12/N] Update scatter with CPU/CUDA implementations
* #86407 [11/N] Update all_to_all with CPU/CUDA implementations
* **#86368 [10/N] Update barrier with CPU/CUDA implementations**
* #86166 [9/N] [Dispatchable Collectives] Update reduce_scatter with CPU / CUDA implementations

### Changes
- Updates for the barrier collective
- NOTE: current change will not achieve dispatching of barrier since there is no tensor to read from

### Context
https://github.com/pytorch/pytorch/issues/86225

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @kwen2501 @awgu

H-Huang committed:
[10/N] Update barrier with CPU/CUDA implementations

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40145698).

awgu commented on torch/csrc/distributed/c10d/OpsImpl.cpp:
```diff
@@ -286,6 +302,15 @@ TORCH_LIBRARY_IMPL(c10d, CPU, m) {
 }
+
+TORCH_LIBRARY_IMPL(c10d, CPU, m) {
+  m.impl("barrier", barrier_cpu);
```
How do you decide if there should be a trailing underscore?

H-Huang replied:
the convention for PT operators (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml) is if the tensor is modified inplace, then operator should be appended with `_`. We don't do this for `barrier` and `send`

awgu replied:
Thank you Professor Huang!

H-Huang committed:
Update on "[10/N] Update barrier with CPU/CUDA implementations"


### Changes
- Updates for the barrier collective

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D40145698](https://our.internmc.facebook.com/intern/diff/D40145698)

[ghstack-poisoned]

H-Huang committed:
Update on "[10/N] Update barrier with CPU/CUDA implementations"


### Changes
- Updates for the barrier collective

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D40145698](https://our.internmc.facebook.com/intern/diff/D40145698)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40145698).

kwen2501 commented on test/distributed/test_c10d_common.py:
```diff
@@ -1488,6 +1490,7 @@ def _test_collectives(self, backend):
             (dist.all_reduce,),
```
Sorry, I may have missed this earlier. Would passing `self.rank` to reduce and broadcast cause each rank identifying a different root and hang? They should have the same root.

H-Huang replied:
Passing self.rank causes the broadcast operation to be sourced from that rank, but it does not hang waiting for other ranks to ACK the broadcast it sent. The same logic applies to reduce so thats why I believe it is not hanging

kwen2501 commented on test/distributed/test_c10d_common.py:
```diff
@@ -1464,7 +1464,9 @@ def _call_collective_with_varying_tensors(self, backend, collective, *args):
+        if collective == dist.barrier:
+            collective()
+        elif collective == dist.all_gather:
             collective([tensor], tensor, *args)
```
1. nit: this if-else block is getting bigger (as a result of wrapping/templating).
Maybe the test would be easier to read if we just write out each test call in _test_collectives, like:
```
dist.barrier()
dist.all_reduce(tensor)
...
```
2. Is `collective([tensor], tensor, *args)` a correct format for `all_gather`? i.e. it will have a list of only **one** tensor for the output. Or are we testing the dispatching functionality with WORLD_SIZE=1 here? (if so the code makes sense)

H-Huang replied:
1. Makes sense, the if statement is getting a bit messy, I will change it in future PRs.
2. We are just testing dispatching functionality and making sure the operation is callable. So handling the latter case, since collective correctness should be handled by other tests.

kwen2501 reviewed (approved):
It is okay to land current changes since they lead to the same behavior.
But we should note that current change would not achieve dispatching of barrier.
Also added some minor comments about the test. (Sorry I missed them previously; can be improved in a future PR)

H-Huang committed:
Update on "[10/N] Update barrier with CPU/CUDA implementations"


### Changes
- Updates for the barrier collective

### Context
https://github.com/pytorch/pytorch/issues/86225

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @kwen2501 @awgu

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge
